### PR TITLE
Add com.dz4k.FruitCredits

### PR DESCRIPTION
**Pull request by dz4k**
### Please confirm your submission meets all the criteria

- [X] Please describe the application briefly.
  **Fruit Credits is a double-entry accounting app for tracking one's personal finances. It is a GUI wrapper around the command-line tool hledger.**

- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project. If not, I contacted upstream developers about this submission. **Link:**

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

File: com.dz4k.FruitCredits.json [hunk 1]
```diff
@@ -0,0 +1,3397 @@
...
+        "mkdir -p /app/lib",
+        "ln -s /usr/lib/x86_64-linux-gnu/libtinfo.so /app/lib/libtinfo.so.5",
+        "./configure --prefix=/app",
+        "make install"
+      ],
+      "sources": [
+        {
+          "sha256": "ff5b4929a4e89c536e7badb3b142e353dc4bda1f31d3ee446406ad88c3bebddf",
+          "type": "archive",
+          "url": "https://downloads.haskell.org/ghc/9.6.6/ghc-9.6.6-x86_64-deb9-linux.tar.xz"
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:74]: how will this work on aarch64?

File: com.dz4k.FruitCredits.json [hunk 2]
```diff
@@ -0,0 +1,3397 @@
...
+      "buildsystem": "meson",
+      "sources": [
+        {
+          "type": "dir",
+          "path": "."
+        },
+        {
+          "type": "git",
+          "url": "https://gitlab.gnome.org/BZHDeveloper/vul.git",
+          "branch": "master",
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:3385]: Never a git branch. You need a specific commit.

File: com.dz4k.FruitCredits.json [hunk 3]
```diff
@@ -0,0 +1,3397 @@
...
+          "dest": "subprojects/vul"
+        },
+        {
+          "type": "git",
+          "url": "https://gitlab.gnome.org/jwestman/blueprint-compiler.git",
+          "branch": "main",
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:3391]: Never a git branch. You need a specific commit.

Also blue-print need to be cleanup as it's build only.

File: com.dz4k.FruitCredits.json [hunk 4]
```diff
@@ -0,0 +1,3397 @@
...
+      ]
+    },
+    {
+      "name": "fruit-credits",
+      "builddir": true,
+      "buildsystem": "meson",
+      "sources": [
+        {
+          "type": "dir",
+          "path": "."
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:3380]: so how does this reference the sources?

File: com.dz4k.FruitCredits.json [hunk 5]
```diff
@@ -0,0 +1,3397 @@
...
+    "--share=ipc",
+    "--socket=fallback-x11",
+    "--device=dri",
+    "--socket=wayland",
+    "--env=LD_LIBRARY_PATH=/app/lib:/app/lib64:/usr/lib/x86_64-linux-gnu/GL/default/lib:/usr/lib/x86_64-linux-gnu/openh264/extra:/usr/lib/x86_64-linux-gnu/openh264/extra:/usr/lib/sdk/vala/lib:/usr/lib/x86_64-linux-gnu/GL/default/lib"
+  ],
+  "build-options": {
+    "append-path": "/usr/lib/sdk/vala/bin",
+    "prepend-ld-library-path": "/usr/lib/sdk/vala/lib",
+    "env": {}
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:21]:
```suggestion
    "prepend-ld-library-path": "/usr/lib/sdk/vala/lib"
```

File: com.dz4k.FruitCredits.json [hunk 6]
```diff
@@ -0,0 +1,3397 @@
...
+    "org.freedesktop.Sdk.Extension.vala"
+  ],
+  "command": "fruit-credits",
+  "finish-args": [
+    "--share=network",
+    "--share=ipc",
+    "--socket=fallback-x11",
+    "--device=dri",
+    "--socket=wayland",
+    "--env=LD_LIBRARY_PATH=/app/lib:/app/lib64:/usr/lib/x86_64-linux-gnu/GL/default/lib:/usr/lib/x86_64-linux-gnu/openh264/extra:/usr/lib/x86_64-linux-gnu/openh264/extra:/usr/lib/sdk/vala/lib:/usr/lib/x86_64-linux-gnu/GL/default/lib"
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:16]: I am pretty sure you don't need this.

File: com.dz4k.FruitCredits.json [hunk 7]
```diff
@@ -0,0 +1,3397 @@
+{
+  "id": "com.dz4k.FruitCredits",
+  "runtime": "org.gnome.Platform",
+  "runtime-version": "47",
+  "sdk": "org.gnome.Sdk",
+  "sdk-extensions": [
+    "org.freedesktop.Sdk.Extension.vala"
```
Comment (hfiguiere) [inline com.dz4k.FruitCredits.json:7]: why do you need this ? valac is in the runtime. If you really need vala.

File: com.dz4k.FruitCredits.json [hunk 8]
```diff
@@ -0,0 +1,3397 @@
...
+        "/bin/hp2ps",
+        "/bin/hpc",
+        "/bin/hsc2hs",
+        "/lib/libtinfo.so.5",
+        "/lib/debug",
+        "/lib/ghc-9.6.6",
+        "/share/man",
+        "/share/doc/ghc-9.6.6"
+      ],
+      "buildsystem": "simple",
```
Comment (bbhtt) [inline com.dz4k.FruitCredits.json:62]: use autotools, put the `mkdir` and `ln -s` steps as `post-install`, remove the other two.

File: com.dz4k.FruitCredits.json [hunk 9]
```diff
@@ -0,0 +1,3397 @@
...
+    },
+    {
+      "name": "data-fix",
+      "buildsystem": "simple",
+      "builddir": false,
+      "build-commands": [
+        "echo '#! /usr/bin/env runhaskell' >Setup.txt",
+        "echo '> import Distribution.Simple' >>Setup.txt",
+        "echo '> main = defaultMain' >>Setup.txt",
+        "ln -s Setup.txt Setup.lhs || true",
```
Comment (bbhtt) [inline com.dz4k.FruitCredits.json:464]: why are these needed?

File: com.dz4k.FruitCredits.json (hunk 2 above)
Comment (dz4k) [reply com.dz4k.FruitCredits.json]: Thanks and fixed.

File: com.dz4k.FruitCredits.json (hunk 3 above)
Comment (dz4k) [reply com.dz4k.FruitCredits.json]: Fixed the commit.

blueprint-compiler is being built and used by meson internally, I haven't found any blueprint-compiler related files in the resulting flatpak unless I'm missing something.

File: com.dz4k.FruitCredits.json (hunk 4 above)
Comment (dz4k) [reply com.dz4k.FruitCredits.json]: It doesn't, I'd changed it temporarily to use my local worktree and forgot to change it back. Fixed now.

File: com.dz4k.FruitCredits.json (hunk 9 above)
Comment (dz4k) [reply com.dz4k.FruitCredits.json:501]: The haskell modules in the manifest are based on the output of [cabal-flatpak](https://hackage.haskell.org/package/cabal-flatpak-0.0) and the article [How to Flatpak a Haskell App into Flathub](https://medium.com/@lettier/how-to-flatpak-a-haskell-app-into-flathub-86ef6d69e94d). Are there more up-to-date resources for bundling Haskell programs?

Comment (bbhtt) [reply com.dz4k.FruitCredits.json:501]: No, we don't have haskell apps. This is probably the second or third one ever.

Comment (hfiguiere) [reply com.dz4k.FruitCredits.json:501]: and the part that is autogenerated should be in its own file to make it easier to maintain.

Comment (dz4k) [reply com.dz4k.FruitCredits.json]: Split the generated modules out to a new file.

File: com.dz4k.FruitCredits.json (hunk 1 above)
Comment (dz4k) [reply com.dz4k.FruitCredits.json]: It won't, sorry. The tool I used to generate the haskell modules has a bug relating to aarch64, which I didn't initially notice. I've added an aarch64 ghc module now.

File: hledger-dep.json [hunk 10]
```diff
@@ -0,0 +1,3386 @@
...
+        "/share/man",
+        "/share/doc/ghc-9.6.6"
+      ],
+      "buildsystem": "simple",
+      "builddir": false,
+      "build-commands": [
+        "mkdir -p /app/lib",
+        "ln -s /usr/lib/x86_64-linux-gnu/libtinfo.so /app/lib/libtinfo.so.5",
+        "./configure --prefix=/app",
+        "make install"
```
Comment (bbhtt) [inline hledger-dep.json:80]: this was not fixed.

File: hledger-dep.json [hunk 11]
```diff
@@ -0,0 +1,3386 @@
...
+        "/lib/debug",
+        "/lib/ghc-9.6.6",
+        "/share/man",
+        "/share/doc/ghc-9.6.6"
+      ],
+      "buildsystem": "simple",
+      "builddir": false,
+      "build-commands": [
+        "mkdir -p /app/lib",
+        "ln -s /usr/lib/x86_64-linux-gnu/libtinfo.so /app/lib/libtinfo.so.5",
```
Comment (bbhtt) [inline hledger-dep.json:78]: this won't work on aarch64

File: hledger-dep.json
Comment (bbhtt) [inline hledger-dep.json:1]: Are you autogenerating this somehow? Can you document the process in a readme?

File: hledger-dep.json (hunk 11 above)
Comment (dz4k) [reply hledger-dep.json:78]: this ghc module is `"only-arches": ["x86-64"]`. the aarch64 version is the one above. i've built the app locally with `--arch=aarch64` and it worked, though i don't have a real aarch64 device to test on

Comment (bbhtt) [reply hledger-dep.json:78]: Ah I see it now

File: hledger-dep.json
Comment (dz4k) [reply hledger-dep.json:1]: it's generated using the tool `cabal-flatpak` and [this shell script](https://codeberg.org/dz4k/fruit-credits/src/branch/main/scripts/build-flathub-manifest.sh). I can move the script to this repo if that would be preferable. And yes I can write a readme too

File: hledger-dep.json (hunk 10 above)
Comment (dz4k) [reply hledger-dep.json:80]: applying your suggestion (as I naively understood it) caused the build to fail with an opaque message, and I was debugging that.

what I did was change
```json
"buildsystem": "simple",
"build-commands": [
  "mkdir -p /app/lib",
  "ln -s /usr/lib/aarch64-linux-gnu/libtinfo.so /app/lib/libtinfo.so.5",
  "./configure --prefix=/app",
  "make install"
],
```
to
```json
"buildsystem": "autotools",
"post-install": [
  "mkdir -p /app/lib",
  "ln -s /usr/lib/aarch64-linux-gnu/libtinfo.so /app/lib/libtinfo.so.5"
],
```
and I got the error
```
...
****************************************************
Configuration done, ready to 'make install'
(see README and INSTALL files for more info.)
****************************************************
Run "make install" to install
make: *** [Makefile:10: default] Error 1
Error: module ghc-9.6.6-aarch64: Child process exited with code 2
```

Comment (dz4k) [reply hledger-dep.json]: It's fixed now.

**Comment by bbhtt** [comment]
bot, build com.dz4k.FruitCredits

**Comment by dz4k** [comment]
bot, build com.dz4k.FruitCredits

**Comment by dz4k** [comment]
```
{
    "errors": [
        "appstream-failed-validation"
    ],
    "appstream": [
        "E: com.dz4k.FruitCredits:11: tag-duplicated summary (lang=tr)"
    ],
    "info": [
        "appstream-failed-validation: Metainfo file /tmp/tmpgjxe8bnz/metainfo/com.dz4k.FruitCredits.metainfo.xml has failed validation. Please see the errors in appstream block"
    ],
    "message": "Please consult the documentation at https://docs.flathub.org/docs/for-app-authors/linter"
}
```
I'm confused... How am I supposed to translate the summary?

https://docs.flathub.org/docs/for-app-authors/metainfo-guidelines/#metainfo-translations

**Comment by bbhtt** [comment]
why do you have static translations in metainfo like that? You are already using the i18n in metainfo https://codeberg.org/dz4k/fruit-credits/src/commit/81cb6b032283a702074f612fccc6ff348e656461/data/meson.build#L15

**Comment by dz4k** [comment]
bot, build com.dz4k.FruitCredits

**Comment by dz4k** [comment]
bot, build com.dz4k.FruitCredits

**Comment by bbhtt** [comment]
Ctrl + N -> filling up date/description and clicking save does not work:


```
** (fruit-credits:2): CRITICAL **: 10:22:25.419: file ../src/hledger/hledger.vala: line 155: uncaught error: Operation not supported (g-io-error-quark, 15)

(fruit-credits:2): GLib-GIO-CRITICAL **: 10:22:25.419: GTask 0x5635a938d690 (source object: 0x5635a4e2a290, source tag: (nil)) finalized without ever returning (using g_task_return_*()). This potentially indicates a bug in the program.
```

**Comment by bbhtt** [comment]
Also where do I get a hledger journal file from?

**Comment by dz4k** [comment]
> Ctrl + N -> filling up date/description and clicking save does not work:

I can't reproduce this, on an empty file or in my personal journal. What file did you use? Did you only fill the date and description, or did you add some accounts and amounts as well?

> Also where do I get a hledger journal file from?

https://github.com/simonmichael/hledger/tree/master/examples has quite a lot, I've been using [this one](https://raw.githubusercontent.com/simonmichael/hledger/refs/heads/master/examples/Cody.journal) as a sample.

**Comment by bbhtt** [comment]
> I can't reproduce this, on an empty file or in my personal journal. What file did you use? Did you only fill the date and description, or did you add some accounts and amounts as well?

Just did Ctrl + N, didn't use anything else.

**Comment by dz4k** [comment]
I see. The "add transaction" feature doesn't work without a journal file (this is by design), but the dialog can still be activated and there's no prompt or other affordance to create a new journal file (this is a defect).

Is this a blocker for merging? I'm happy to set up my own repo for the few people who wanted to try the app and reopen this when the app is more polished.

**Comment by bbhtt** [comment]
No it's fine if it was not meant to be used that way, but you should add a dialoge blocking add transaction without any journal loaded.
